### PR TITLE
Added composite key provider for storing composite keys in keystore

### PR DESCRIPTION
**Pull request by patrickkuo**
* Moved Composite Key related classes to crypto.composite
* New CordaSecurityProvider to provide composite key mappings for Java security.
* Added new Composite Key type to Crypto
* Test case for testing storing and retrieving composite key to/from keystore.

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/crypto/composite/CompositeKey.kt [hunk 1]
```diff
@@ -103,17 +108,16 @@ class CompositeKey private constructor (val threshold: Int,
...
-            require (weight > 0) { "A non-positive weight was detected. Node info: $this"  }
+            require(weight > 0) { "A non-positive weight was detected. Node info: $this" }
         }
+
         override fun compareTo(other: NodeAndWeight): Int {
             if (weight == other.weight) {
                 return node.hashCode().compareTo(other.node.hashCode())
-            }
-            else return weight.compareTo(other.weight)
+            } else return weight.compareTo(other.weight)
```
Comment (rnicoll) [inline core/src/main/kotlin/net/corda/core/crypto/composite/CompositeKey.kt]: Can you take the brackets off the first return, and put this second one on its own line, please? Shouldn't be much longer, but will be more consistent and easier to follow the flow (IMHO).

Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/crypto/composite/CompositeKey.kt]: @patrickkuo as I merged a new version of compareTo, do the ^ after resolving conflicts.

File: core/src/main/kotlin/net/corda/core/crypto/provider/CordaSecurityProvider.kt [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+    init {
+        AccessController.doPrivileged(PrivilegedAction<Unit> { setup() })
+    }
+
+    private fun setup() {
+        put("KeyFactory.${CompositeKey.KEY_ALGORITHM}", "net.corda.core.crypto.composite.KeyFactory")
+        put("Signature.${CompositeSignature.SIGNATURE_ALGORITHM}", "net.corda.core.crypto.composite.CompositeSignature")
+
+        val compositeKeyOID = CompositeSignature.SIGNATURE_ALGORITHM_IDENTIFIER.algorithm.id
```
Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/crypto/provider/CordaSecurityProvider.kt]: Why do we use the same OID for both composite keys and signatures?

Comment (patrickkuo) [inline core/src/main/kotlin/net/corda/core/crypto/provider/CordaSecurityProvider.kt]: I am following EdDSA's convention... it has a signature algorithm name `NONEwithEdDSA` for EdDSA but there are no OID for that.

Comment (patrickkuo) [inline core/src/main/kotlin/net/corda/core/crypto/provider/CordaSecurityProvider.kt]: Ah you are right, `NONEwithEdDSA` probably have it's own OID, just like our `CompositeSigniture`, I will make one up for `CompositeKey`

File: core/src/main/kotlin/net/corda/core/crypto/Crypto.kt [hunk 3]
```diff
@@ -684,7 +685,7 @@ object Crypto {
         val pointQ = FixedPointCombMultiplier().multiply(parameterSpec.g, deterministicD)
         // This is unlikely to happen, but we should check for point at infinity.
         if (pointQ.isInfinity)
-            // Instead of throwing an exception, we retry with SHA256(seed).
+        // Instead of throwing an exception, we retry with SHA256(seed).
```
Comment (rnicoll) [inline core/src/main/kotlin/net/corda/core/crypto/Crypto.kt]: As this comment is part of the if conditional, it should be indented inline with the line below shouldn't it?

Comment (patrickkuo) [inline core/src/main/kotlin/net/corda/core/crypto/Crypto.kt]: Intellij auto formatted it..... will change it back

File: core/src/test/kotlin/net/corda/core/crypto/CompositeKeyTests.kt [hunk 4]
```diff
@@ -76,6 +87,17 @@ class CompositeKeyTests {
     }
 
     @Test
+    fun `der encoded tree decodes correctly`() {
```
Comment (kchalkias) [inline core/src/test/kotlin/net/corda/core/crypto/CompositeKeyTests.kt]: For this particular test, let's better add weights in some of the nodes (different to 1).

**Pull request merged**